File: LeetCode/Python/078 Subsets.py
# -*- coding: utf-8 -*-

'''
Subsets
=======

Given a set of distinct integers, nums, return all possible subsets.

Note:
Elements in a subset must be in non-descending order.
The solution set must not contain duplicate subsets.
For example,
If nums = [1,2,3], a solution is:

[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
'''


class Solution(object):
    '''算法思路：

    利用规律：nums 的每一位分别对应 bits（长度为len(nums)）的每一位
    '''
    def subsets(self, nums):
        nums.sort()

        r = []
        for i in xrange(2 ** len(nums)):
            subset, n = [], 0
            while i:
                if i & 1:
                    subset.append(nums[n])
                i >>= 1
                n += 1
            r.append(subset)

        return r


s = Solution()
print s.subsets([2, 3, 4, 5])

File: LeetCode/C++/078. Subsets.cc
# include "header.h"


class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> r;
        sort(nums.begin(), nums.end());
        auto n = int(pow(2, nums.size()));

        for (auto i = 0; i < n; ++i) {
            vector<int> subset;
            auto x = i, j = 0;

            while (x) {
                if (x & 1) subset.push_back(nums[j]);
                ++j;
                x >>= 1;
            }
            r.push_back(subset);
        }
        return r;
    }
};


int main() {
    Solution s;
    vector<int> v = {1, 2, 3};
    auto r = s.subsets(v);

    for (auto i = r.begin(); i != r.end(); ++i) {
        auto array = *i;
        for (auto j = array.begin(); j != array.end(); ++j) {
            cout << *j << endl;
        }
        cout << "------------------" << endl;
    }
}
